File: problems/Number of Connected Components in an Undirected Graph/countComponents.cs
public class Solution {
    public int CountComponents(int n, int[][] edges) {
        Unions dsu = new Unions(n);

        foreach (var edge in edges)
        {
            dsu.Union(edge[0], edge[1]);
        }

        ISet<int> roots = new HashSet<int>();
        for (int i = 0; i < n; i++)
        {
            roots.Add(dsu.Find(i));
        }

        return roots.Count;
    }

    private class Unions
    {
        private readonly int[] _parents;
        private readonly int[] _ranks;

        public Unions(int n)
        {
            _parents = new int[n];
            _ranks = new int[n];
            for (int i = 0; i < n; i++)
            {
                _parents[i] = i;
            }
        }

        public int Find(int x)
        {
            if (x != _parents[x])
            {
                x = Find(_parents[x]);
            }
            return _parents[x];
        }

        public bool Union(int x, int y)
        {
            int px = Find(x);
            int py = Find(y);
            if (px == py)
            {
                return false;
            }
            if (_ranks[px] > _ranks[py])
            {
                _parents[py] = px;
                _ranks[px]++;
            }
            else
            {
                _parents[px] = py;
                _ranks[py]++;
            }
            return true;
        }
    }
}

File: problems/Number of Connected Components in an Undirected Graph/countComponents.java
class Solution {
    public int countComponents(int n, int[][] edges) {
        HashSet<Integer> store = new HashSet<>();
        Dsu dsu = new Dsu(n);

        for (int[] edge : edges) {
            dsu.union(edge[0], edge[1]);
        }

        for (int idx = 0; n > idx; ++idx) {
            store.add(dsu.find(idx));
        }

        return store.size();
    }

    class Dsu {
        int[] parents;
        int[] ranks;

        public Dsu(int n) {
            parents = new int[n];
            ranks = new int[n];

            for (int idx = 0; n > idx; ++idx) {
                parents[idx] = idx;
            }
        }

        public int find(int a) {
            if (a != parents[a]) {
                parents[a] = find(parents[a]);
            }
            return parents[a];
        }

        public void union(int a, int b) {
            int pa = find(a);
            int pb = find(b);

            if (pa != pb) {
                if (ranks[pa] > ranks[pb]) {
                    parents[pa] = pb;
                    ++ranks[pa];
                } else {
                    parents[pb] = pa;
                    ++ranks[pb];
                }
            }
        }
    }
}
